cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceTransformExecutorFactory.java

```diff
@@ -101,11 +107,94 @@ protected TrackedTransform getTransformation(String pluginType, String stageName
                                                                            mapOutputValClassName),
                                        stageMetrics);
       }
+    } else if (BatchJoiner.PLUGIN_TYPE.equals(pluginType)) {
+      BatchJoiner<?, ?, ?> batchJoiner = pluginInstantiator.newPluginInstance(stageName);
+      BatchRuntimeContext runtimeContext = createRuntimeContext(stageName);
+      batchJoiner.initialize(runtimeContext);
+      StageMetrics stageMetrics = new DefaultStageMetrics(metrics, stageName);
+      if (isMapper) {
+        return getTrackedJoinOnStep(new MapperJoinerTransformation(batchJoiner, mapOutputKeyClassName), stageMetrics);
+      } else {
+        return getTrackedMergeStep(new ReducerJoinerTransformation(batchJoiner, mapOutputKeyClassName), stageMetrics);
+      }
     }
     return super.getTransformation(pluginType, stageName);
   }
 
   /**
+   * A Transformation that uses an join's joinOn method. Converts join value to tagged map output with stage name for
+   * reducer. It uses {@link MapTaggedOutputWritable} to tag join value with stage name so that we can use this
+   * in mapreduce.
+   *
+   * @param <JOIN_KEY> type of join key
+   * @param <JOIN_VALUE> type of the join value
+   * @param <OUT> type of the output of joiner
+   * @param <OUT_KEY> type of the map output
+   */
+  private static class MapperJoinerTransformation<JOIN_KEY, JOIN_VALUE, OUT, OUT_KEY extends Writable>
+    implements Transformation<JOIN_VALUE, KeyValue<OUT_KEY, MapTaggedOutputWritable<JOIN_VALUE>>> {
+    private final Joiner<JOIN_KEY, JOIN_VALUE, OUT> joiner;
+    private final WritableConversion<JOIN_KEY, OUT_KEY> keyConversion;
+
+    MapperJoinerTransformation(Joiner<JOIN_KEY, JOIN_VALUE, OUT> joiner, String joinKeyClassName) {
+      this.joiner = joiner;
+      WritableConversion<JOIN_KEY, OUT_KEY> keyConversion = WritableConversions.getConversion(joinKeyClassName);
+      this.keyConversion = keyConversion == null ? new CastConversion<JOIN_KEY, OUT_KEY>() : keyConversion;
+    }
+
+    @Override
+    public void transform(JOIN_VALUE input, Emitter<KeyValue<OUT_KEY, MapTaggedOutputWritable<JOIN_VALUE>>> emitter)
+      throws Exception {
+      String stageName;
+      // TODO Use input to get stageName
+      if (emitter instanceof TransformDetail) {
+        stageName = ((TransformDetail) emitter).getPrevStage();
+      } else {
+        stageName = ((TransformDetail) ((TrackedEmitter) emitter).getEmitter()).getPrevStage();
```

Comment:
Like the TODO comment mentions, it seems like the input should carry both the stage that emitted it, as well as the value emitted. I'm not sure how this approach works, unless it relies on the fact that the transform dag is called in a depth first manner?